### PR TITLE
[spec] New instructions

### PR DESCRIPTION
**Pull request by rossberg**
Add the (non-controversial ones of the) new instructions to the spec.

This is the same as #50, but rebased on master.

@conrad-watt, PTAL. For some strange reason, GH doesn't let me add you as a reviewer this time.

### CODE REVIEW

**Comment by rossberg** [comment]
@conrad-watt, friendly ping.

**Comment by conrad-watt** [comment]
on this now, sorry

File: document/core/appendix/gen-index-instructions.py [hunk 1]
```diff
@@ -269,13 +269,13 @@ def Instruction(name, opcode, type=None, validation=None, execution=None, operat
...
-    Instruction(r'\REFNULL~t', r'\hex{D0}', r'[] \to [t]', r'valid-ref.null', r'exec-ref.null'),
-    Instruction(r'\REFISNULL', r'\hex{D1}', r'[t] \to [\I32]', r'valid-ref.is_null', r'exec-ref.is_null'),
+    Instruction(r'\REFNULL~\X{ht}', r'\hex{D0}', r'[] \to [(\REF~\NULL~\X{ht})]', r'valid-ref.null', r'exec-ref.null'),
+    Instruction(r'\REFISNULL', r'\hex{D1}', r'[(\REF~\NULL~\X{ht})] \to [\I32]', r'valid-ref.is_null', r'exec-ref.is_null'),
     Instruction(r'\REFFUNC~x', r'\hex{D2}', r'[] \to [\FUNCREF]', r'valid-ref.func', r'exec-ref.func'),
-    Instruction(None, r'\hex{D3}'),
-    Instruction(None, r'\hex{D4}'),
+    Instruction(r'\REFASNONNULL', r'\hex{D3}', r'[(\REF~\NULL~\X{ht})] \to [(\REF~\X{ht})]', r'valid-ref.as_non_null', r'exec-ref.as_non_null'),
+    Instruction(r'\BRONNULL~l', r'\hex{D4}', r'[t^\ast~(\REF~\NULL~\X{ht})] \to [t^\ast~(\REF~\X{ht})]', r'valid-br_on_null', r'exec-br_on_null'),
     Instruction(None, r'\hex{D5}'),
```
Comment (conrad-watt) [inline document/core/appendix/gen-index-instructions.py:277]: Is there a reason `br_on_null` isn't next to `br_on_non_null` in the index space?

File: document/core/syntax/instructions.rst [hunk 2]
```diff
@@ -188,11 +189,13 @@ Instructions in this group are concerned with accessing :ref:`references <syntax
    \production{instruction} & \instr &::=&
      \dots \\&&|&
      \REFNULL~\reftype \\&&|&
```
Comment (conrad-watt) [inline document/core/syntax/instructions.rst:191]: `\reftype` here should be `\heaptype` - sorry if I missed this previously.

File: document/core/appendix/gen-index-instructions.py (hunk 1 above)
Comment (rossberg) [reply document/core/appendix/gen-index-instructions.py:277]: D5 was already assigned to ref.eq at the time we added br_on_non_null.

File: document/core/syntax/instructions.rst (hunk 2 above)
Comment (rossberg) [reply document/core/syntax/instructions.rst:191]: Fixed.

**Review by rossberg — COMMENTED** [review]
Thanks for the review.

**Review by conrad-watt — APPROVED** [review]
LGTM

**Pull request merged**